### PR TITLE
Bugfix/3396 wrong 404 redirect

### LINKED ISSUE #3396: 404 and infinite redirect on category pages in multistore mode
## Current behavior

If I got to a category page in a non-default store like `http://www.test.de/de/boys.html` the page will return a 404 (in multistore mode). But it will work in the default store without store-code in url like `http://www.test.de/de/boys.html`.

I first tracked it down to the changes for the new 404 page from #3296 & #3243, when the `beforeEachGuard` in [`core/modules/url/router/beforeEach.ts`](https://github.com/DivanteLtd/vue-storefront/blob/119be33841f1b5eec650d6e7764fc2c1fdf1c236/core/modules/url/router/beforeEach.ts#L30) is asking for `to.matched[0].name === 'page-not-found'` but the route name in multistore mode will be like `de-page-not-found`. So I changed it to `to.matched[0].name.endsWith('page-not-found')` to support multistores at this point.

Now it will serve the correct category page but will end in an infinite redirect in the browser (client-side) floating the memory.

The next thing I noticed is that the matched `dynamicRoutes` of `UrlDispatchMapper` are have url paths like `/uk/de/boys.html` after some infinite loops, so I looked at the `processDynamicRoute` method in [`core/modules/url/helpers/index.ts`](https://github.com/DivanteLtd/vue-storefront/blob/119be33841f1b5eec650d6e7764fc2c1fdf1c236/core/modules/url/helpers/index.ts#L18-L42) and fixed this by removing the store-code from path using `removeStoreCodeFromRoute` like:
```javascript
import { removeStoreCodeFromRoute } from '@vue-storefront/core/lib/multistore'
…

export function processDynamicRoute (routeData: LocalizedRoute, fullPath: string, addToRoutes: boolean = true): LocalizedRoute[] {
  fullPath = removeStoreCodeFromRoute(fullPath) as string
  ...
}
```
(I have mentioned this in https://github.com/DivanteLtd/vue-storefront/issues/3359#issuecomment-523325285 before – probably at the wrong place)

Now the url paths in the `dynamicRoutes` is correct, but there still is this infinite redirect on the client-side. It is always calling the `de-page-not-found` route, again and again. The server-side response and code is delivered correctly.

… I will try to track this down a little further and will keep you updated – but If you have any hints or already an idea what could cause this, please let me know.

## Expected behavior

Should also show category page with store-code in URL path.

## Steps to reproduce the issue

1. Set `storeViews.multistore` to `true`
1. Go to specific store category page like `http://localhost:3000/de/women/tops-women/jackets-women/jackets-23`
1. It should return a 404 response
1. If you go to the same page without a store-code like `http://localhost:3000/women/tops-women/jackets-women/jackets-23`, it should work

## Repository

develop (since 3d2b4a2e69b8e0707f7ecd58c320fb0172a27271)

## Can you handle fixing this bug by yourself?

- [ ] YES (I try, and will let you know if I could)
- [x] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [x] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details

- Browser: Chrome
- OS: OSX
- Node: 10 LTS
- Code Version: 1.11.0-rc.1 - develop 3d2b4a2e69b8e0707f7ecd58c320fb0172a27271

### PR DESCRIPTION
**Pull request by lukeromanowicz**
### Related issues
closes #3396

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and I am aware that every new feature should be a module

### CODE REVIEW

**Comment by ResuBaka** [comment]
For me all the symptoms are gone. The flashing is gone and I can't reproduce the redirect loop with appending the store code.

**Comment by pkarw** [comment]
@ResuBaka @cewald @lukeromanowicz just one thing - maybe this is the good moment to add unit tests for url module? :-)

**Comment by ResuBaka** [comment]
Yea you are right we should at least write some for the functions which have changed here.

@lukeromanowicz or @cewald do you want to write the tests other wise I can to it?

**Comment by cewald** [comment]
Good idea. I'm sadly out of time until next week.
Maybe put it into another issue to not loose track of it?

**Comment by lukeromanowicz** [comment]
I am out of time as well. Maybe we could merge it and add a note to the issue to add the tests `later.`. It can stay open until tests will get added.

File: core/modules/url/router/beforeEach.ts [hunk 1]
```diff
@@ -36,12 +36,21 @@ export async function beforeEachGuard (to: Route, from: Route, next) {
   const fullPath = normalizeUrlPath(to.fullPath)
   const hasRouteParams = to.hasOwnProperty('params') && Object.values(to.params).length > 0
   const isPreviouslyDispatchedDynamicRoute = to.matched.length > 0 && to.name && to.name.startsWith('urldispatcher')
-  if (!to.matched.length || to.matched[0].name === 'page-not-found' || (isPreviouslyDispatchedDynamicRoute && !hasRouteParams)) {
+
+  if (!to.matched.length || to.matched[0].name.endsWith('page-not-found') || (isPreviouslyDispatchedDynamicRoute && !hasRouteParams)) {
```
Comment (patzick) [inline core/modules/url/router/beforeEach.ts]:
```suggestion
  if (!to.matched.length || to.matched[0].name.includes('page-not-found') || (isPreviouslyDispatchedDynamicRoute && !hasRouteParams)) {
```

File: core/lib/router-manager.ts [hunk 2]
```diff
@@ -10,8 +10,10 @@ const RouterManager = {
         (registeredRoute) => registeredRoute.name === route.name && registeredRoute.path === route.path
       ) === -1
     )
-    this._registeredRoutes.push(...uniqueRoutes)
-    router.addRoutes(uniqueRoutes)
+    if (uniqueRoutes.length > 0) {
```
Comment (patzick) [inline core/lib/router-manager.ts]:
```suggestion
    if (uniqueRoutes.length) {
```

**Pull request merged**